src/vivarium/framework/resource.py

```diff
@@ -11,147 +11,294 @@
 NULL_RESOURCE_TYPE = 'null'
 
 
-class ResourceProducer:
+class ResourceGroup:
+    """Resource groups are the nodes in the resource dependency graph.
+
+    A resource group represents the pool of resources produced by a single
+    callable and all the dependencies necessary to produce that resource.
+    When thinking of the dependency graph, this represents a vertex and
+    all in-edges.  This is a local-information representation that can be
+    used to construct the entire dependency graph once all resources are
+    specified.
+
+    """
 
     def __init__(self, resource_type: str, resource_names: List[str], producer: MethodType, dependencies: List[str]):
-        self.resource_type = resource_type
-        self.resource_names = resource_names
-        self.producer = producer
-        self.dependencies = dependencies
+        self._resource_type = resource_type
+        self._resource_names = resource_names
+        self._producer = producer
+        self._dependencies = dependencies
 
-    def __repr__(self):
-        resources = ', '.join([f'{self.resource_type}.{name}' for name in self.resource_names])
-        return f'ResourceProducer({resources})'
+    @property
+    def type(self) -> str:
+        """The type of resource produced by this resource group's producer.
 
-    def __str__(self):
-        resources = ', '.join([f'{self.resource_type}.{name}' for name in self.resource_names])
-        return f'({resources})'
+        Must be one of :data:`RESOURCE_TYPES`.
 
+        """
+        return self._resource_type
 
-class EmptySet:
+    @property
+    def names(self) -> List[str]:
+        """The long names (including type) of all resources in this group."""
+        return [f'{self._resource_type}.{name}' for name in self._resource_names]
 
-    def add(self, item: Any):
-        pass
+    @property
+    def producer(self) -> Any:
+        """The method or object that produces this group of resources."""
+        return self._producer
 
-    def __contains__(self, item: Any) -> bool:
-        return False
+    @property
+    def dependencies(self) -> List[str]:
+        """The long names (including type) of dependencies for this group."""
+        return self._dependencies
 
+    def __iter__(self) -> Iterable[str]:
+        return iter(self.names)
 
-class ResourceGroup:
+    def __repr__(self) -> str:
+        resources = ', '.join(self)
+        return f'ResourceProducer({resources})'
 
-    def __init__(self, phase: str):
-        self.phase = phase
-        self.resources = {}
-        self._graph = None
-        # null producers are those that don't produce any resources externally but still consume
-        # other resources (i.e., have dependencies) - these are only pop initializers as of 9/26/2019
+    def __str__(self) -> str:
+        resources = ', '.join(self)
+        return f'({resources})'
+
+
+class ResourceManager:
+    """Manages all the resources needed for population initialization."""
```

Comment:
You can try to grok the diff, or read this as a reimplementation of the original.  The resource manager had almost no responsibilites besides creating resource groups, so I gutted that and moved all functionality to the manager.